src/goto-symex/execution_state.cpp

```diff
@@ -838,7 +838,8 @@ void execution_statet::get_expr_globals(
       name == "c:@__ESBMC_is_dynamic" ||
       name == "c:@__ESBMC_blocked_threads_count" ||
       name.find("c:pthread_lib") != std::string::npos ||
-      name == "c:@__ESBMC_rounding_mode")
+      name == "c:@__ESBMC_rounding_mode" ||
```

Comment:
I wonder whether there is a more efficient way to perform these string comparisons, something like:

````CPP
#include <unordered_set>

// Precompute the set of strings to compare against
static const std::unordered_set<std::string> excludedNames = {
    "c:@__ESBMC_alloc", "c:@__ESBMC_alloc_size",
    "c:@__ESBMC_is_dynamic", "c:@__ESBMC_blocked_threads_count",
    "c:@__ESBMC_rounding_mode"
};

// Check if the given name is excluded
bool isExcluded(const std::string& name) {
    return excludedNames.count(name) > 0 ||
           name.find("c:pthread_lib") != std::string::npos;
}

// Usage example:
if (isExcluded(name)) {
    return;
}
````

Reply:
I doubt it. The string equality-based comparisons check the length first.

I guess it's more performant to check the common prefix `__ESBMC_` and then do string comparisons. The latter will usually be quick since the strings have different lengths and/or start with a different symbol.

But benchmarks will tell the truth...

Reply:
It should be faster to compare the dstring instead of the string, i.e., the thename from the symbol, instead of calling as_string()

Reply:
The approaches have little preformance difference in esbmc-unix and esbmc-unix2:

this PR + hashset: 489.3s + 717.5s
check common prefix first + hashset: 487.7s + 713.9s
dstring + hashset: 488.9s + 720.5s.

And there could be deviation because each one is single run on my laptop.
How about further evaulate this performance in other PR?

Reply:
> How about further evaulate this performance in other PR?

Sure, that's fine. As a clarification, my comment was recommending against any sort of hash table for these comparisons, because hash table lookups require hashing the entire string before the first array indexing is performed and then checking equality of the strings obtained in that bucket (if any) - the hashing cost makes sense to pay when the number of strings in the table is large in comparison to the lengths of the strings being looked up.